### PR TITLE
Introduce missing service statuses

### PR DESCRIPTION
**Pull request by pierlo-upitup**
This PR introduces support for the `DELAYED` and `WAITING` status. From the [Marathon docs](https://mesosphere.github.io/marathon/docs/marathon-ui.html#application-status-reference):

> Additionally, by inspecting the v2/queue endpoint the following states can also be determined:

**Waiting**

Marathon is waiting for offers from Mesos. True whenever an app has `queueEntry.delay.overdue === true`.

**Delayed**

An app is considered delayed whenever too many tasks of the application failed in a short amount of time. Marathon will pause this deployment and retry later. True if the queue endpoint returns the following JSON condition: `queueEntry.delay.overdue === false`

image

TODO (out of scope)
- [ ] I'd like some guidance on how to add fixtures for the integration tests covering these new statuses in the `SidebarFilter-cy.js` test case
- [ ] verify progress bar styles match the new status values

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @daiyiwhale

File: src/js/constants/ServiceStatusTypes.js [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
-  RUNNING: 0,
-  DEPLOYING: 1,
-  SUSPENDED: 2,
-  NA: 4
+  NA: 0,
+  RUNNING: 1,
+  DEPLOYING: 2,
+  SUSPENDED: 3,
+  DELAYED: 4,
+  WAITING: 5
```
Comment (orlandohohmeier) [inline src/js/constants/ServiceStatusTypes.js]: I'm not so sure about the order... Isn't "delayed" more important than "waiting" and "running"/"deploying" more important than "suspended"?

Comment (pierlo-upitup) [inline src/js/constants/ServiceStatusTypes.js]: So you suggest:

```
NA
SUSPEND
RUNNING
DEPLOYING
WAITING
DELAYED
```
?

File: src/js/stores/DCOSStore.js [hunk 2]
```diff
@@ -98,6 +113,21 @@ class DCOSStore extends EventEmitter {
...
+  onMarathonQueueChange(nextQueue) {
+    let {marathon:{queue}} = this.data;
+
+    nextQueue.forEach((entry) => {
+      if (entry.app == null) {
+        return;
+      }
+
+      let serviceId = entry.app.id;
+      queue.set(serviceId, entry);
```
Comment (orlandohohmeier) [inline src/js/stores/DCOSStore.js]: You don't need the `serviceId` variable here...

Comment (pierlo-upitup) [inline src/js/stores/DCOSStore.js]: no but I thought it made the code easier to read

File: src/js/structs/Service.js [hunk 3]
```diff
@@ -126,6 +126,15 @@ module.exports = class Service extends Item {
   getServiceStatus() {
     let {tasksRunning} = this.getTasksSummary();
     let deployments = this.getDeployments();
+    let queue = this.getQueue();
+
+    if (queue != null && queue.delay && queue.delay.overdue) {
```
Comment (orlandohohmeier) [inline src/js/structs/Service.js]: I'd like some documentation here, as it's really hard to distinguish the both conditions.

File: src/js/structs/ServiceTree.js [hunk 4]
```diff
@@ -168,21 +169,40 @@ module.exports = class ServiceTree extends Tree {
   }
 
   getServiceStatus() {
-    let {tasksRunning} = this.getTasksSummary();
-    let deployments = this.getDeployments();
+    // Return the service status with the highest index
+    let treeStatus = this.getStatusSummary().reduce((memo, count, statusKey) => {
+      if (count > 0) {
+        Object.keys(ServiceStatus).forEach(statusId => {
```
Comment (orlandohohmeier) [inline src/js/structs/ServiceTree.js]: Do we really need to iterate through the `ServiceStatus` object over and over again? How about creating a temporary map, or adding a different getter?

File: src/js/constants/ServiceStatusTypes.js (hunk 1 above)
Comment (orlandohohmeier) [inline src/js/constants/ServiceStatusTypes.js]: Yep.

File: src/js/structs/ServiceTree.js [hunk 5]
```diff
@@ -168,21 +169,40 @@ module.exports = class ServiceTree extends Tree {
   }
 
   getServiceStatus() {
-    let {tasksRunning} = this.getTasksSummary();
-    let deployments = this.getDeployments();
+    // Return the service status with the highest index
+    let treeStatus = this.getStatusSummary().reduce((memo, count, statusKey) => {
```
Comment (orlandohohmeier) [inline src/js/structs/ServiceTree.js]: Wy are you reducing the `getStatusSummary` result?  And what's the difference?

File: src/js/stores/DCOSStore.js [hunk 6]
```diff
@@ -71,6 +79,13 @@ class DCOSStore extends EventEmitter {
   }
 
   /**
+   * Fetch launch queue from Marathon
+   */
+  fetchQueue() {
```
Comment (aldipower) [inline src/js/stores/DCOSStore.js]: I would prefer a more descriptive name here, so we could get rid of the distracting comment.
Maybe `fetchMarathonLaunchQueue()`.
As always just a suggestion.

Comment (pierlo-upitup) [inline src/js/stores/DCOSStore.js]: how about `fetchServicesQueue` ?

File: src/js/structs/Service.js (hunk 3 above)
Comment (aldipower) [inline src/js/structs/Service.js]: No documentation please for such things (that nobody reads).

To make it more descriptive we could refactor the code, which makes it even faster:
```JavaScript
if (queue != null && queue.delay) {
  if (queue.delay.overdue) {
     return ServiceStatus.WAITING;
  }
  return ServiceStatus.DELAYED;
}
```

File: src/js/stores/DCOSStore.js (hunk 6 above)
Comment (aldipower) [inline src/js/stores/DCOSStore.js]: Sounds good.

File: src/js/structs/Service.js (hunk 3 above)
Comment (pierlo-upitup) [inline src/js/structs/Service.js]: This would be such a perfect place for a ternary, 😢

File: src/js/structs/ServiceTree.js (hunk 5 above)
Comment (pierlo-upitup) [inline src/js/structs/ServiceTree.js]: I realise the code is not very terse. I will think of a more self-explanatory way of doing the same thing.

File: src/js/stores/DCOSStore.js (hunk 6 above)
Comment (pierlo-upitup) [inline src/js/stores/DCOSStore.js]: LOL i just realised we dont even need this!

File: src/js/structs/ServiceTree.js [hunk 7]
```diff
@@ -168,21 +168,19 @@ module.exports = class ServiceTree extends Tree {
...
-    let deployments = this.getDeployments();
-
-    if (deployments.length > 0) {
-      return ServiceStatus.DEPLOYING;
-    }
-
-    if (tasksRunning > 0) {
-      return ServiceStatus.RUNNING;
-    }
+    return this.reduceItems(function (statusSummary, item) {
```
Comment (orlandohohmeier) [inline src/js/structs/ServiceTree.js]: I'd rename the parameter to something as simple as `serviceStatus`, as we not  dealing with summary anymore.  #nit

Comment (pierlo-upitup) [inline src/js/structs/ServiceTree.js]: Good point. I'll go for `serviceTreeStatus` though, as it represents the final tree status.

File: src/js/stores/DCOSStore.js (hunk 2 above)
Comment (orlandohohmeier) [inline src/js/stores/DCOSStore.js]: I'm ok with that and would even encourage us to do this more often, but I think we agreed that we not store properties (for what ever reason) if we only access it once. Your call.

**Comment by orlandohohmeier** [comment]
Great job! ✨ 💯

**Comment by pierlo-upitup** [comment]
All comments addressed.

File: src/js/stores/DCOSStore.js [hunk 8]
```diff
@@ -218,7 +240,10 @@ class DCOSStore extends EventEmitter {
         return item;
       }
 
-      let options = {versions: versions.get(item.getId())};
+      let options = {
+        versions: versions.get(item.getId()),
+        queue: queue.get(item.getId())
```
Comment (aldipower) [inline src/js/stores/DCOSStore.js]: Second call of the getId-function, please cache it.

**Comment by aldipower** [comment]
Cool! :) Works and looks good!

**Pull request merged**